### PR TITLE
CPU optimized inference

### PR DESCRIPTION
**Pull request by taolei87**

### CODE REVIEW

**Comment by hpasapp** [comment]
can we announce this in #ml-common too please? Seems like conceptually this is not dissimilar to #ml-common, in that it is code used by both mleng and research?

File: sru/sru_cpu_impl.cpp [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+std::vector<at::Tensor> cpu_forward(
+        const at::Tensor & U,
+        const at::Tensor & x,
+        const at::Tensor & weight_c,
+        const at::Tensor & bias,
+        const at::Tensor & c_init,
+        const at::Tensor & mask_pad,
+        int64_t length, 
+        int64_t batch_size, 
+        int64_t d,
```
Comment (hpasapp) [inline sru/sru_cpu_impl.cpp]: what is d? what is k? can we either have a comment saying what they are, or else give them more descriptive names please?

File: sru/sru_cpu_impl.cpp [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+
+float sigmoidf(float x);
+float reluf(float x);
+float seluf(float x);
+float apply_activation(int64_t type, float x);
+
+
+/* Implementation starts here */
+
+#define CHECK_CONTIGUOUS(x) AT_ASSERTM(x.is_contiguous(), #x " must be contiguous")
```
Comment (hpasapp) [inline sru/sru_cpu_impl.cpp]: these macros seem nice

File: sru/sru_cpu_impl.cpp [hunk 3]
```diff
@@ -0,0 +1,240 @@
+
+#include <torch/torch.h>
+#include <vector>
+
+std::vector<at::Tensor> cpu_forward(
```
Comment (hpasapp) [inline sru/sru_cpu_impl.cpp]: part of me thinks that declarations belong in a header file; most of me thinks that clearly what you have currently works, and is probably simpler to build; so I'm good with keeping them here for now.

File: sru/sru_cpu_impl.cpp [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+                int l_ = (j < d) ? l : (length - l - 1);
+                if (pad_ptr && (*(pad_ptr + l_*batch_size))) continue;
+
+                // U[l_,i,j,*]
+                const float* u = U_ptr + l_*ncols_u + (i*d2+j)*k;
+                const float u0 = *(u);
+                const float u1 = *(u+1);
+                const float u2 = *(u+2);
+
+                const float fw = *(forget_w_ptr + j);
```
Comment (hpasapp) [inline sru/sru_cpu_impl.cpp]: just out of curiosity, why not write eg `forget_w_ptr[j]`, `reset_w_ptr[j]`, etc ?

**Comment by hpasapp** [comment]
looks good to me. I think it'd be good to announce this in #ml-common on the whole, in case anyone else wants to review.

File: sru/sru_cpu_impl.cpp [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+
+                // forget gate
+                const float fg = sigmoidf(u1 + fw*prev_c + fb);
+                // reset gate
+                const float rg = sigmoidf(u2 + rw*prev_c + rb);
+
+                const float c_val = (prev_c - u0) * fg + u0; // prev_c * fg + u0 * (1 - fg);
+                const float gc_val = apply_activation(activation_type, c_val);
+                
+                const float x_val = has_skip_term ? ((k == 4) ? u[3] : x_ptr[l_*ncols + offset]*scale_x) : 0;
```
[the remark this answers is not in the document]
Comment (taoleicn) [reply sru/sru_cpu_impl.cpp]: yeah. the `has_skip_term` and `k` are constants so I guess the branch prediction should work very well here.

File: sru/sru_cpu_impl.cpp [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+                const float rw = reset_w_ptr[j];
+                const float fb = forget_b_ptr[j];
+                const float rb = reset_b_ptr[j];
+
+                const float prev_c = c_ptr[offset];
+
+                // forget gate
+                const float fg = sigmoidf(u1 + fw*prev_c + fb);
+                // reset gate
+                const float rg = sigmoidf(u2 + rw*prev_c + rb);
```
[the remark this answers is not in the document]
Comment (taoleicn) [reply sru/sru_cpu_impl.cpp]: i can test it a bit.

**Comment by cdfox-asapp** [comment]
After the change I just committed, confirmed that I can `pip install .` and then do autosuggest training, run a server with the trained model, and send requests to it. Really fast! 42ms for 300 length contexts (this is the upper limit of what we'd see with current configuration used in production).

**Pull request merged**